### ISSUE TITLE
weird gpu memory usage

### ISSUE THREAD

**Issue by dwSun** [apache/incubator-mxnet#11243]
## Description
The script below will eat all gpu memory gradually in the first epoch. Sometimes it will cause OOM error and crash.
Sometimes it just keep running, but at the 2rd epoch, the gpu memory and cuda core usage will drop to a very low level suddenly.

## Environment info (Required)
ubuntu 18.04 cuda91
nvidia driver 390.48
nvidia 1080ti gpu card

mxnet-cu91 1.2.0 from pypi (with or without mkl)

## Minimum reproducible example
```py
#!/usr/bin/env python3
import time
from multiprocessing import cpu_count

import mxnet as mx
from mxnet import autograd, gluon
from mxnet.gluon.data import dataset
from mxnet.gluon.data.vision import transforms
from mxnet.gluon.model_zoo import vision

CPU_COUNT = cpu_count()
mx.random.seed(42)  # set seed for repeatability

class ImgDataSet(dataset.Dataset):
    def __getitem__(self, idx):
        return mx.nd.random.uniform(shape=(320, 320, 3)), 0

    def __len__(self):
        return 50000

transformer = transforms.Compose([
    transforms.Resize(224),
    transforms.ToTensor()])


train_dataset = ImgDataSet()
train_dataset = train_dataset.transform_first(transformer)

batch_size = 16
train_data_loader = mx.gluon.data.DataLoader(
    train_dataset, batch_size, shuffle=True, num_workers=CPU_COUNT)

# construct and initialize network.
ctx = mx.gpu() if mx.test_utils.list_gpus() else mx.cpu()

net = vision.get_mobilenet_v2(1.0, ctx=ctx, classes=1000)
net.hybridize()
net.initialize(mx.init.Xavier(factor_type='out'), ctx=ctx)
# define loss and trainer.
criterion = gluon.loss.SoftmaxCrossEntropyLoss()
trainer = gluon.Trainer(net.collect_params(), 'Adam', {
    'learning_rate': 0.1, 'wd': 4e-5})

epochs = 50
for epoch in range(epochs):
    total_train_loss = mx.nd.zeros(1, ctx=ctx)
    start = time.time()
    training_samples = 0
    for batch_idx, (data, label) in enumerate(train_data_loader):
        data = data.as_in_context(ctx)
        label = label.as_in_context(ctx)

        with autograd.record():
            output = net(data)
            loss = criterion(output, label)
        loss.backward()
        trainer.step(data.shape[0])
        total_train_loss += loss.sum()
        training_samples += data.shape[0]
        if batch_idx > 0 and batch_idx % 500 == 0:
            elapse = time.time() - start
            # uncomment this line will prevent this error to happen.
            # print(total_train_loss.asscalar()/training_samples)
            print('{}:[{}] samples/s in train'.format(batch_idx,
                                                      training_samples/elapse))

```
If uncomment the asscalar line, the code run well but become slower and slower.
```
0.014223459476006483
500:[310.3343436613137] samples/s in train
0.0071204068301083685
1000:[138.02850671172925] samples/s in train
0.004750276866712068
1500:[117.94254639226769] samples/s in train
0.0035655089821117272
2000:[109.79963955911568] samples/s in train
0.002856015301093918
2500:[105.23927686823332] samples/s in train
0.0023851065745317153
3000:[102.53342911171562] samples/s in train
0.0020499249861193397
3500:[100.58626881090342] samples/s in train
0.0017987247467994452
4000:[99.1741771153156] samples/s in train
0.0016033576875494682
4500:[98.11294445964649] samples/s in train
0.0014470566084231885
5000:[97.28087273484665] samples/s in train
0.0013191592929537048
5500:[96.63322991265628] samples/s in train
0.0012125678905505494
6000:[95.973076875728] samples/s in train
3.9965115086047233e-05
500:[138.14346868877303] samples/s in train
3.9967515698441495e-05
1000:[110.0120848330557] samples/s in train
3.996974652723659e-05
1500:[102.58358700612278] samples/s in train

```

**Comment by kalyc** [comment]
Hello @dwSun, thanks for submitting your issue. There is an open discussion forum about mxnet on discuss.mxnet.io - you could post your query there to learn more.

**Comment by ThomasDelteil** [comment]
@dwSun MXNet is built using asynchronous operations. When you load data, run a forward pass or backward pass, the operations are enqueued on the MXNet backend and executed when the parent dependencies are available. 

With your current script, there is no blocking operations, so the training runs through your epoch and keep adding "copy to GPU" operations. These operations don't have parent dependencies and can be executed immediately.  the actual training isn't completed when you reach the end of your epoch loop. After a few epoch you will clog up your GPU memory.

`print(total_train_loss.asscalar()/training_samples)`

`.asscalar()` is `.asnumpy()[0]`, this causes a synchronous operation to copy the memory back to the CPU. When you add this line, the training isn't "slow" it is the normal speed, because every 500 iterations, your network is going to wait for the computation to be completed and return the result to the CPU. If your dataset is small and fit in GPU memory, you can have a `mx.nd.waitall` at the end of your epoch, but that means your entire dataset will be copied to GPU. This makes it pretty fast since at the beginning of each batch the data is already available. However you might run into OOM error, in that case you can for example keep track of your loss using `loss_acc += loss.sum().asscalar()` and forcing a copy to CPU on every batch.

**Comment by dwSun** [comment]
@ThomasDelteil, thanks for explain that.
